src/source/PeerConnection/Rtp.c

```diff
@@ -122,27 +122,47 @@ STATUS transceiverOnBandwidthEstimation(PRtcRtpTransceiver pRtcRtpTransceiver, U
     return retStatus;
 }
 
-UINT64 convertTimestampToRTP(UINT64 clockRate, UINT64 pts)
-{
-    return (pts * clockRate) / HUNDREDS_OF_NANOS_IN_A_SECOND;
-}
-
-STATUS writeFrame(PRtcRtpTransceiver pRtcRtpTransceiver, PFrame pFrame)
+STATUS writeFrame(PRtcRtpTransceiver pRtcRtpTransceiver, PRtcFrame pRtcFrame)
 {
     STATUS retStatus = STATUS_SUCCESS;
     PKvsPeerConnection pKvsPeerConnection = NULL;
     PKvsRtpTransceiver pKvsRtpTransceiver = (PKvsRtpTransceiver) pRtcRtpTransceiver;
     BOOL locked = FALSE, bufferAfterEncrypt = FALSE;
     PRtpPacket pPacketList = NULL, pRtpPacket = NULL;
-    UINT32 i = 0, packetLen = 0, allocSize;
+    UINT32 i = 0, packetLen = 0, headerLen = 0, allocSize;
     PBYTE rawPacket = NULL;
     PPayloadArray pPayloadArray = NULL;
     RtpPayloadFunc rtpPayloadFunc = NULL;
+    UINT64 randomRtpTimeoffset = 0; // TODO: spec requires random rtp time offset
     UINT64 rtpTimestamp = 0;
+    UINT64 now = GETTIME();
+    PFrame pFrame = (PFrame) pRtcFrame;
 
     CHK(pKvsRtpTransceiver != NULL, STATUS_NULL_ARG);
     pKvsPeerConnection = pKvsRtpTransceiver->pKvsPeerConnection;
     pPayloadArray = &(pKvsRtpTransceiver->sender.payloadArray);
+    ATOMIC_ADD(&pKvsRtpTransceiver->sender.outboundRtpStreamStats.totalEncodedBytesTarget, pFrame->size);
+    ATOMIC_ADD(&pKvsRtpTransceiver->sender.outboundRtpStreamStats.totalEncodeTime, pRtcFrame->encodeTimeMsec);
+    ATOMIC_STORE(&pKvsRtpTransceiver->sender.outboundRtpStreamStats.targetBitrate, pRtcFrame->targetBitrate);
+    if (MEDIA_STREAM_TRACK_KIND_VIDEO == pKvsRtpTransceiver->sender.track.kind) {
+        ATOMIC_INCREMENT(&pKvsRtpTransceiver->sender.outboundRtpStreamStats.framesEncoded);
+        if (0 != (pFrame->flags & FRAME_FLAG_KEY_FRAME)) {
+            ATOMIC_INCREMENT(&pKvsRtpTransceiver->sender.outboundRtpStreamStats.keyFramesEncoded);
+        }
+        if (pKvsRtpTransceiver->sender.lastKnownFrameCountTime == 0) {
+            pKvsRtpTransceiver->sender.lastKnownFrameCountTime = now;
+            pKvsRtpTransceiver->sender.lastKnownFrameCount     = pKvsRtpTransceiver->sender.outboundRtpStreamStats.framesEncoded;
+        } else if (now - pKvsRtpTransceiver->sender.lastKnownFrameCountTime > HUNDREDS_OF_NANOS_IN_A_SECOND) {
+            UINT64 frames = pKvsRtpTransceiver->sender.outboundRtpStreamStats.framesEncoded - pKvsRtpTransceiver->sender.lastKnownFrameCount;
```

Comment:
Btw, I have not set clang checks with aligned alignments. So, this would break.

Reply:
ok